### PR TITLE
Eventhubs: Retry Implementation and Test

### PR DESCRIPTION
**Pull request by mssfang**
This PR includes:

(1) Implementation of Retry, RetryExponential
(2) Unit test of implementation of them.

fixes: #3751

### CODE REVIEW

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import com.azure.eventhubs.implementation.ClientConstants;
+
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public abstract class Retry {
+
+    public static final Retry NO_RETRY = new RetryExponential(Duration.ofSeconds(0), Duration.ofSeconds(0), 0);
+
+    private AtomicInteger retryCount = new AtomicInteger(0);
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: `private final` rather than just `private`. I like to make it clear that this field variable should only be set once.

Comment (conniey) [reply eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: You can just do `new AtomicInteger()`, it will set the initial count to 0.

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+import com.azure.core.amqp.exception.AmqpException;
+import com.azure.core.amqp.exception.ErrorCondition;
+import com.azure.eventhubs.implementation.ClientConstants;
+
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public abstract class Retry {
+
+    public static final Retry NO_RETRY = new RetryExponential(Duration.ofSeconds(0), Duration.ofSeconds(0), 0);
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: Use `Duration.ZERO`

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+     * @param exception A exception that was observed for the operation to be retried.
+     * @return true if the exception is a retryable exception, otherwise false.
+     * @throws IllegalArgumentException when the exception is null.
+     */
+    public static boolean isRetryableException(Exception exception) {
+        if (exception == null) {
+            throw new IllegalArgumentException("exception cannot be null");
+        }
+
+        if (exception instanceof AmqpException) {
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: You can simplify this entire block by going:
`return (exception instanceof AmqpException) && ((AmqpException) exception).isTransient()`

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        }
+        return false;
+    }
+
+    /**
+     * Get 'NO_RETRY' of current.
+     *
+     * @return Retry 'NO_RETRY'.
+     */
+    public static Retry getNoRetry() {
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: Retry.NO_RETRY is already public, this getter is not necessary.

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    /**
+     * Get default configured Retry.
+     *
+     * @return Retry which has all default property set up.
+     */
+    public static Retry getDefault() {
+        return new RetryExponential(
+            ClientConstants.DEFAULT_RETRY_MIN_BACKOFF,
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: Pull all the ClientConstants into this class. ClientConstants is a giant bag of random values, which makes it hard to reason about. If we pull these three constants into this class, we know that they definitely pertain to Retry.

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+     * Calculates the amount of time to delay before the next retry attempt
+     *
+     * @param lastException The last exception that was observed for the operation to be retried.
+     * @param remainingTime The amount of time remaining for the cumulative timeout across retry attempts.
+     * @return The amount of time to delay before retrying the associated operation; if {@code null}, then the operation is no longer eligible to be retried.
+     */
+    public Duration getNextRetryInterval(Exception lastException, Duration remainingTime) {
+        int baseWaitTime = 0;
+
+        if (lastException == null || !(lastException instanceof AmqpException)) {
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: If the lastException is null or not an instance of AmqpException, we should return `null`. For example, an IllegalArgumentException is not retriable.

Comment (conniey) [reply eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: You should also test that the exception is retryable or not, and if they have hit the maximum number of attempts. These are checks that are agnostic of the retry algorithm.

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/RetryExponential.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Override
+    protected Duration onGetNextRetryInterval(final Exception lastException,
+                                              final Duration remainingTime,
+                                              final int baseWaitSeconds,
+                                              final int retryCount) {
+        if ((!Retry.isRetryableException(lastException)) || retryCount >= maxRetryCount) {
+            return null;
+        }
+        double nextRetryInterval = Math.pow(retryFactor, (double) retryCount);
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/RetryExponential.java]: In general, if you know a variable won't be reassigned, I'd add `final` to it. This makes it explicit that you don't want this to change. So if, someone makes a mistake down the line and tries to reassign it, it'll result in a compiler error.

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/RetryExponential.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+        long nextRetryIntervalSeconds = (long) nextRetryInterval;
+        long nextRetryIntervalNano = (long) ((nextRetryInterval - (double) nextRetryIntervalSeconds) * 1000000000);
+        if (remainingTime.getSeconds() < Math.max(nextRetryInterval, ClientConstants.TIMER_TOLERANCE.getSeconds())) {
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/RetryExponential.java]: Similarly, move the ClientConstant into this class.

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+
+public abstract class Retry {
+
+    public static final Retry NO_RETRY = new RetryExponential(Duration.ofSeconds(0), Duration.ofSeconds(0), 0);
+
+    private AtomicInteger retryCount = new AtomicInteger(0);
+
+    /**
+     * Check if the existing exception is a retryable exception.
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: The spelling is "retriable" rather than "retryable". I noticed the difference in our SDK vs the others.

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java (hunk 6 above)
Comment (mssfang) [reply eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: both tests are tested in test cases: 
(1) isRetriable: https://github.com/Azure/azure-sdk-for-java/pull/3818/files#diff-88f19a64404192ec0b032ec6f648b628R57
(2) Excess max retry:  https://github.com/Azure/azure-sdk-for-java/pull/3818/files#diff-88f19a64404192ec0b032ec6f648b628R72

Comment (conniey) [reply eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: What I mean is that the logic for if the exception is retryable or not and max attempts (including maxRetryCount) should be in Retry.java rather than ExponentialRetry.java because the max retries and is transient applies to _all_ retry policies.

Comment (mssfang) [reply eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: fixed

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/ExponentialRetry.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    protected Duration calculateNextRetryInterval(final Exception lastException,
+                                              final Duration remainingTime,
+                                              final int baseWaitSeconds,
+                                              final int retryCount) {
+        if ((!Retry.isRetriableException(lastException)) || retryCount >= super.maxRetryCount()) {
+            return null;
+        }
+        final double nextRetryInterval = Math.pow(retryFactor, (double) retryCount);
+
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/ExponentialRetry.java]: nit: random blank line. I try to keep my variable declarations together, leave a blank line and then start invoking methods, etc on them.

File: eventhubs/client/azure-eventhubs/src/test/java/com/azure/eventhubs/RetryTest.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+        retry.resetRetryInterval();
+        Assert.assertEquals(0, retry.getRetryCount());
+
+        retry.incrementRetryCount();
+        Assert.assertEquals(1, retry.getRetryCount());
+    }
+
+    @Test
+    public void isRetryable() {
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/test/java/com/azure/eventhubs/RetryTest.java]: nit: isRetriable. and notRetriable.

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    /**
+     * Get default configured Retry.
+     *
+     * @return Retry which has all default property set up.
+     */
+    public static Retry getDefaultRetry() {
+        return new ExponentialRetry(
+            DEFAULT_RETRY_MIN_BACKOFF,
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/Retry.java]: nit: Any reason each parameter is on a new line? Does it reach the 120 characters/line limit?

File: eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/ExponentialRetry.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+import java.time.Duration;
+
+/**
+ * A policy to govern retrying of messaging operations in which the delay between retries
+ * will grow in an exponential manner, allowing more time to recover as the number of retries increases.
+ */
+public final class ExponentialRetry extends Retry {
+
+    public static final Duration TIMER_TOLERANCE = Duration.ofSeconds(1);
```
Comment (conniey) [inline eventhubs/client/azure-eventhubs/src/main/java/com/azure/eventhubs/ExponentialRetry.java]: nit: add a blank line between `public static` and `private final` variables. It makes it easier to visually group the different types of variables.
